internal/pkg/addon/package.go

```diff
@@ -273,8 +281,129 @@ type s3Object struct {
 }
 
 func (a *Addons) uploadAddonAsset(assetPath string, forceZip bool) (s3Object, error) {
+	// make path absolute from wsPath
+	if !path.IsAbs(assetPath) {
+		assetPath = path.Join(a.wsPath, assetPath)
+	}
+
+	info, err := a.fs.Stat(assetPath)
+	if err != nil {
+		return s3Object{}, err
+	}
+
+	var asset asset
+	if forceZip || info.IsDir() {
+		asset, err = a.zipAsset(assetPath)
+	} else {
+		asset, err = a.fileAsset(assetPath)
+	}
+	if err != nil {
+		return s3Object{}, fmt.Errorf("create asset: %w", err)
+	}
+
+	s3Path := artifactpath.AddonAsset(a.wlName, asset.hash)
+
+	url, err := a.uploader.Upload(a.bucket, s3Path, asset.data)
+	if err != nil {
+		return s3Object{}, fmt.Errorf("upload %s to s3 bucket %s: %w", assetPath, a.bucket, err)
+	}
+
+	bucket, key, err := s3.ParseURL(url)
+	if err != nil {
+		return s3Object{}, fmt.Errorf("parse s3 url: %w", err)
+	}
+
+	fmt.Printf("Uploaded %s to s3 at: %s\n", assetPath, s3.Location(bucket, key))
 	return s3Object{
-		Bucket: a.bucket,
-		Key:    "TODO",
+		Bucket: bucket,
+		Key:    key,
+	}, nil
+}
+
+type asset struct {
+	data io.Reader
+	hash string
+}
+
+// zipAsset creates an asset from the directory or file specified by root
+// where the data is the compressed zip archive, and the hash is
+// a hash of each files name, permission, and content.
+func (a *Addons) zipAsset(root string) (asset, error) {
+	buf := &bytes.Buffer{}
+	z := zip.NewWriter(buf)
+	defer z.Close()
+
+	hash := sha256.New()
+
+	if err := a.fs.Walk(root, func(path string, info fs.FileInfo, err error) error {
+		switch {
+		case err != nil:
+			return err
+		case info.IsDir():
+			return nil
+		}
+
+		fname, err := filepath.Rel(root, path)
+		switch {
+		case err != nil:
+			return fmt.Errorf("rel: %w", err)
+		case fname == ".":
+			fname = info.Name()
+		}
+
+		f, err := a.fs.Open(path)
+		if err != nil {
+			return fmt.Errorf("open: %w", err)
+		}
+		defer f.Close()
+
+		header, err := zip.FileInfoHeader(info)
+		if err != nil {
+			return fmt.Errorf("create zip file header: %w", err)
+		}
+
+		header.Name = fname
+		header.Method = zip.Deflate
```

Comment:
👍 nice! For my edification, do we this because now the file is going to be written to the zip archive? like I wonder why the `CreateHeader` method doesn't do this by default internally

Reply:
which one are you asking about? `zip.Deflate`? maybe this answers your question 😊 
> FileInfoHeader creates a partially-populated FileHeader from an fs.FileInfo. Because fs.FileInfo's Name method returns only the base name of the file it describes, it may be necessary to modify the Name field of the returned header to provide the full path name of the file. If compression is desired, callers should set the FileHeader.Method field; it is unset by default.